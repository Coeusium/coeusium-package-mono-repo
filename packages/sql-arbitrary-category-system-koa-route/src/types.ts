export type {
  ConnectionInfo,
  CategorySet,
  Category,
} from 'sql-arbitrary-category-system';
