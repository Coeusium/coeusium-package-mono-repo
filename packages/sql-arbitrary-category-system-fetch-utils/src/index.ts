export * from './sql-arbitrary-category-system-fetch-utils';
export type { Category, CategorySet } from './types';
