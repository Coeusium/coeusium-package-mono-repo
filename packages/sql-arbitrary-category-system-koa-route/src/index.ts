export * from './sql-arbitrary-category-system-koa-route';
export type { ConnectionInfo, Category, CategorySet } from './types';
