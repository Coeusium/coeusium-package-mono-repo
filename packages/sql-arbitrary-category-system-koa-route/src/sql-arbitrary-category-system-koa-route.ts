import Router from '@koa/router';
import compose from 'koa-compose';
import { validateQueryParamsAll } from '@almaclaine/koa-utils';
import { ConnectionInfo } from './types';
import {
  addCategory,
  addCategorySet,
  categoryIdExists,
  categorySetIdExists,
  deleteCategory,
  deleteCategorySet,
  getCategory,
  getCategoryChildren,
  getCategoryDescendants,
  getCategorySet,
  getTopLevelCategories,
  listCategorySets,
} from 'sql-arbitrary-category-system';

export { destroy, setupCategorySystem } from 'sql-arbitrary-category-system';

// Category Set Handlers

export function createGetCategorySetHandler(dbInfo: ConnectionInfo) {
  return async (ctx, next) => {
    const { id } = ctx.request.query;
    ctx.body = await getCategorySet(dbInfo, id);
  };
}

export function createGetCategorySetListHandler(dbInfo: ConnectionInfo) {
  return async (ctx, next) => {
    const { limit, page } = ctx.request.query;
    ctx.body = await listCategorySets(dbInfo, page, limit);
  };
}

export function createCategorySetGetHandler(dbInfo: ConnectionInfo) {
  const getCategorySetHandler = createGetCategorySetHandler(dbInfo);
  const listCategorySetHandler = createGetCategorySetListHandler(dbInfo);
  return async (ctx, next) => {
    const { id } = ctx.request.query;
    if (id) {
      await getCategorySetHandler(ctx, next);
    } else {
      await listCategorySetHandler(ctx, next);
    }
  };
}

export function createPostCategorySetHandler(dbInfo: ConnectionInfo) {
  return async (ctx, next) => {
    const { name } = ctx.request.query;
    const id = await addCategorySet(dbInfo, name);
    ctx.body = { id };
  };
}

export function createCategorySetDeleteHandler(dbInfo: ConnectionInfo) {
  return async (ctx, next) => {
    const { id } = ctx.request.query;
    const exists = await categorySetIdExists(dbInfo, id);
    if (!exists) {
      ctx.status = 412;
      ctx.body = { error: 'No category set exists with id ' + id };
      return;
    }

    await deleteCategorySet(dbInfo, id);
    ctx.body = {};
  };
}

export function getCategorySetRouter(dbInfo: ConnectionInfo) {
  const router = new Router();

  const tagGetHandler = createCategorySetGetHandler(dbInfo);
  router.get('/', tagGetHandler);

  const tagDeleteHandler = createCategorySetDeleteHandler(dbInfo);
  router.delete(
    '/',
    compose([validateQueryParamsAll(['id']), tagDeleteHandler]),
  );

  const postTagHandler = createPostCategorySetHandler(dbInfo);
  router.post('/', compose([validateQueryParamsAll(['name']), postTagHandler]));

  return router;
}

// Category Handlers

export function createGetCategoryHandler(dbInfo: ConnectionInfo) {
  return async (ctx, next) => {
    const { id } = ctx.request.query;
    ctx.body = await getCategory(dbInfo, id);
  };
}

export function createGetTopLevelCategoriesHandler(dbInfo: ConnectionInfo) {
  return async (ctx, next) => {
    const { id } = ctx.request.query;
    ctx.body = await getTopLevelCategories(dbInfo, id);
  };
}

export function createGetCategoryChildrenHandler(dbInfo: ConnectionInfo) {
  return async (ctx, next) => {
    const { id } = ctx.request.query;
    ctx.body = await getCategoryChildren(dbInfo, id);
  };
}

export function createGetCategoryDescendantsHandler(dbInfo: ConnectionInfo) {
  return async (ctx, next) => {
    const { id } = ctx.request.query;
    ctx.body = await getCategoryDescendants(dbInfo, id);
  };
}

export function createCategoryGetHandler(dbInfo: ConnectionInfo) {
  const getCategoryHandler = createGetCategoryHandler(dbInfo);
  const getCategoryChildrenHandler = createGetCategoryChildrenHandler(dbInfo);
  const getCategoryDescendantsHandler = createGetCategoryDescendantsHandler(
    dbInfo,
  );
  const getTopLevelCategoriesHandler = createGetTopLevelCategoriesHandler(
    dbInfo,
  );
  return async (ctx, next) => {
    const { type } = ctx.request.query;
    if (type === 'children') {
      await getCategoryChildrenHandler(ctx, next);
    } else if (type === 'descendants') {
      await getCategoryDescendantsHandler(ctx, next);
    } else if (type === 'top') {
      await getTopLevelCategoriesHandler(ctx, next);
    } else {
      await getCategoryHandler(ctx, next);
    }
  };
}

export function createPostCategoryHandler(dbInfo: ConnectionInfo) {
  return async (ctx, next) => {
    const { name, set_id, parent } = ctx.request.query;
    const id = await addCategory(dbInfo, name, set_id, parent);
    ctx.body = { id };
  };
}

export function createCategoryDeleteHandler(dbInfo: ConnectionInfo) {
  return async (ctx, next) => {
    const { id } = ctx.request.query;
    const exists = await categoryIdExists(dbInfo, id);
    if (!exists) {
      ctx.status = 412;
      ctx.body = { error: 'No category exists with id ' + id };
      return;
    }

    await deleteCategory(dbInfo, id);
    ctx.body = {};
  };
}

export function getCategoryRouter(dbInfo: ConnectionInfo) {
  const router = new Router();

  const tagGetHandler = createCategoryGetHandler(dbInfo);
  router.get('/', compose([validateQueryParamsAll(['id']), tagGetHandler]));

  const tagDeleteHandler = createCategoryDeleteHandler(dbInfo);
  router.delete(
    '/',
    compose([validateQueryParamsAll(['id']), tagDeleteHandler]),
  );

  const postTagHandler = createPostCategoryHandler(dbInfo);
  router.post(
    '/',
    compose([validateQueryParamsAll(['name', 'set_id']), postTagHandler]),
  );

  return router;
}

export function getScopedRouter(dbInfo: ConnectionInfo) {
  const catRouter = new Router();
  const catRoutes = getCategoryRouter(dbInfo);

  const router = new Router();
  const setRoutes = getCategorySetRouter(dbInfo);
  router.use('/', catRoutes.routes(), catRoutes.allowedMethods());
  router.use('/set', setRoutes.routes(), setRoutes.allowedMethods());

  catRouter.use('/category', router.routes(), router.allowedMethods());

  return catRouter;
}
