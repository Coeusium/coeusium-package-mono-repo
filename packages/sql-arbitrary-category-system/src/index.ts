export * from './sql-arbitrary-category-system';
export type { ConnectionInfo, Category, CategorySet } from './types';
