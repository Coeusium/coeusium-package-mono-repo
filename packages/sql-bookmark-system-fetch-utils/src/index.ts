export * from './sql-bookmark-system-fetch-utils';
export type { Bookmark } from './types';
