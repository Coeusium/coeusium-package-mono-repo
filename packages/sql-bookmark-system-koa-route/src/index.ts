export * from './sql-bookmark-system-koa-route';
export type { ConnectionInfo, Bookmark } from './types';
