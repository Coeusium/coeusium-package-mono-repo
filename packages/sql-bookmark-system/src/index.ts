export * from './sql-bookmark-system';
export type { ConnectionInfo, Bookmark } from './types';
