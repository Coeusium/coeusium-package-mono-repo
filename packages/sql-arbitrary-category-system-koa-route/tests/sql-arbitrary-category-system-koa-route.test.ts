import * as sqlArbitaryTagSystemKoaRoute from '../src';

describe('sql-arbitrary-category-system-koa-route', () => {
  it('Has Correct API', () => {
    const keys = Object.keys(sqlArbitaryTagSystemKoaRoute);
    expect(keys.length).toBe(17);
    expect(keys).toMatchSnapshot();
  });
});
