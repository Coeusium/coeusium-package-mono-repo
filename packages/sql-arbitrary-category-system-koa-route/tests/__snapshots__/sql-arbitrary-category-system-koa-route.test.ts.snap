// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sql-arbitrary-category-system-koa-route Has Correct API 1`] = `
Array [
  "destroy",
  "setupCategorySystem",
  "createGetCategorySetHandler",
  "createGetCategorySetListHandler",
  "createCategorySetGetHandler",
  "createPostCategorySetHandler",
  "createCategorySetDeleteHandler",
  "getCategorySetRouter",
  "createGetCategoryHandler",
  "createGetTopLevelCategoriesHandler",
  "createGetCategoryChildrenHandler",
  "createGetCategoryDescendantsHandler",
  "createCategoryGetHandler",
  "createPostCategoryHandler",
  "createCategoryDeleteHandler",
  "getCategoryRouter",
  "getScopedRouter",
]
`;
